### PR TITLE
Goto instrument loop unwinding

### PR DESCRIPTION
**Pull request by danpoe**
Loop unwinding with:
- partial loops, assume, assert, rest loops
- unwindsets

### CODE REVIEW

**Comment by danpoe** [comment]
The Linux builds now also work with a new travis.yml file that provides a more recent version of libstdc++.

**Comment by kroening** [comment]
May I ask for some reasonable squashing of commits, please?

**Comment by danpoe** [comment]
Ok, done.

File: regression/goto-instrument-unwind/unwind-chain.sh [hunk 1]
```diff
@@ -6,22 +6,18 @@ goto_instrument=$src/goto-instrument/goto-instrument
 cbmc=$src/cbmc/cbmc
 
 function usage() {
-  echo "Usage: chain unwind-num test_file.c"
+  echo "Usage: chain <options> <testfile>"
```
Comment (tautschnig) [inline regression/goto-instrument-unwind/unwind-chain.sh]: If the testfile were required to be the first (instead of the last) argument then the code could be simplified and all the quotes would become unnecessary. More below.

File: regression/goto-instrument-unwind/unwind-chain.sh [hunk 2]
```diff
@@ -6,22 +6,18 @@ goto_instrument=$src/goto-instrument/goto-instrument
...
-echo "Usage: process test_file.c to goto program"
-
-if [ $# -eq 2 ]
-then
-  unwind_num=$1
-  name=`echo $2 | cut -d. -f1`
-  echo $unwind_num
-  echo $name
-else
+if [ $# -ne 2 ]; then
```
Comment (tautschnig) [inline regression/goto-instrument-unwind/unwind-chain.sh]: Use [ -z $# ] if moving the file argument into the first position and add || [ ! -s $1 ] to ensure it's a valid file(name).

File: regression/goto-instrument-unwind/unwind-chain.sh [hunk 3]
```diff
@@ -6,22 +6,18 @@ goto_instrument=$src/goto-instrument/goto-instrument
...
   usage
+  exit 1
 fi
 
+name=$(echo $2 | cut -d. -f1)
```
Comment (tautschnig) [inline regression/goto-instrument-unwind/unwind-chain.sh]: To address both the argument swapping and add support for path names containing "." use:
```
name=$(echo $1 | sed 's/\.[^\.]*$//')
shift
```

File: regression/goto-instrument-unwind/unwind-chain.sh [hunk 4]
```diff
@@ -6,22 +6,18 @@ goto_instrument=$src/goto-instrument/goto-instrument
...
+
 $goto_cc -o $name.gb $name.c
-$goto_instrument  $name.gb ${name}-unwound.gb --unwind $unwind_num
```
Comment (tautschnig) [inline regression/goto-instrument-unwind/unwind-chain.sh]: $unwind_num is actually undefined now. With the `shift` above use:
$goto_instrument $name.gb ${name}-unwound.gb $@

File: src/goto-instrument/goto_instrument_parse_options.cpp [hunk 5]
```diff
@@ -1324,6 +1360,10 @@ void goto_instrument_parse_optionst::help()
     "Semantic transformations:\n"
     " --nondet-volatile            makes reads from volatile variables non-deterministic\n"
     " --unwind <n>                 unwinds the loops <n> times\n"
+    " --unwindset L:B,...          unwind loop L with a bound of B\n"
+    " --partial-loops              permit paths with partial loops\n"
+    " --unwinding-assertions       generate unwinding assertions\n"
+    " --rest-loops                 add loop for remaining iterations after unwound part\n"
```
Comment (tautschnig) [inline src/goto-instrument/goto_instrument_parse_options.cpp]: I find "rest-loops" hard to parse/understand. "remaining" instead of "rest" would seem clearer to me, but there might be other ideas?

File: src/goto-instrument/unwind.cpp [hunk 6]
```diff
@@ -1,18 +1,132 @@
...
 
-Module: loop unwinding
+Module: Loop unwinding
 
 Author: Daniel Kroening, kroening@kroening.com
+        Daniel Poetzl
 
 \*******************************************************************/
 
+#include <stdexcept>
```
Comment (tautschnig) [inline src/goto-instrument/unwind.cpp]: See comments in string_utils.cpp

File: src/goto-instrument/unwind.cpp [hunk 7]
```diff
@@ -1,18 +1,132 @@
...
+#include <iostream>
```
Comment (tautschnig) [inline src/goto-instrument/unwind.cpp]: Is this a reminiscent of debugging?

File: src/goto-instrument/unwind.cpp [hunk 8]
```diff
@@ -1,18 +1,132 @@
...
+
 #include <util/std_expr.h>
 #include <goto-programs/goto_functions.h>
+#include <util/string_utils.h>
```
Comment (tautschnig) [inline src/goto-instrument/unwind.cpp]: Nit picking: collect all util/ in one group, i.e., move it one line up.

File: src/goto-instrument/unwind.cpp [hunk 9]
```diff
@@ -1,18 +1,132 @@
...
+
+  Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+void parse_unwindset(const std::string &us, unwind_sett &unwind_set)
```
Comment (tautschnig) [inline src/goto-instrument/unwind.cpp]: Ideally we would just be doing this once across the codebase.

File: src/goto-instrument/unwind.cpp [hunk 10]
```diff
@@ -23,14 +137,18 @@ Function: unwind
...
-  goto_programt::targett loop_exit,
-  const unsigned k)
+  const goto_programt::const_targett loop_head,
+  const goto_programt::const_targett loop_exit,
+  const unsigned k,
+  const unwind_strategyt unwind_strategy) const
 {
-  std::vector<goto_programt::targett> exit_points;
-  unwind(goto_program, loop_head, loop_exit, k, exit_points);
+  assert(k>=0);
```
Comment (tautschnig) [inline src/goto-instrument/unwind.cpp]: k is unsigned, this is trivially true.

File: src/goto-instrument/unwind.cpp [hunk 11]
```diff
@@ -45,136 +163,254 @@ Function: unwind
...
+  const goto_programt::const_targett loop_exit,
   const unsigned k,
-  std::vector<goto_programt::targett> &iteration_points)
+  const unwind_strategyt unwind_strategy,
+  std::vector<goto_programt::targett> &iteration_points) const
 {
-  assert(k!=0);
-  
-  iteration_points.resize(k);
+  assert(k>=0);
```
Comment (tautschnig) [inline src/goto-instrument/unwind.cpp]: k is unsigned

File: src/goto-instrument/unwind.cpp [hunk 12]
```diff
@@ -45,136 +163,254 @@ Function: unwind
...
 
-  // loop_exit: where to go after the loop ends
-  // loop_iter: where to go for the next iteration
-  
-  // Add a 'goto' and a 'skip' _before_ loop_exit.
-  // The goto is to take care of 'fall-out' loop exit, and is
-  // not needed if there is an unconditional goto before loop_exit.
+  // rest program after unwound part
+  goto_programt rest_program;
+  assert(rest_program.instructions.empty());
```
Comment (tautschnig) [inline src/goto-instrument/unwind.cpp]: That's a fresh goto_programt, should be trivially true?!

File: src/goto-instrument/unwind.cpp [hunk 13]
```diff
@@ -45,136 +163,254 @@ Function: unwind
...
-      assert(t!=goto_program.instructions.end());
-      goto_programt::targett copied_t=copies.add_instruction();
-      *copied_t=*t;
-      target_vector.push_back(copied_t);
+      copies.destructive_append(tmp_program);
     }
+  }
+  else
+  {
+    assert(k==0);
```
Comment (tautschnig) [inline src/goto-instrument/unwind.cpp]: If I read the code correctly, this is the else case of k!=0 - should be trivially true.

File: src/goto-programs/goto_program_template.h [hunk 14]
```diff
@@ -294,6 +300,14 @@ class goto_program_templatet
     return instructions.insert(target, instructiont());
   }
   
+  //! Insertion before the given target
+  //! \return newly inserted location
+  inline targett insert_before(const_targett target)
+  {
+    return instructions.insert(target, instructiont());
+  }
+
```
Comment (tautschnig) [inline src/goto-programs/goto_program_template.h]: Nit: just one blank line

File: regression/goto-instrument-unwind/unwind-chain.sh (hunk 1 above)
Comment (danpoe) [inline regression/goto-instrument-unwind/unwind-chain.sh]: Doesn't test.pl always pass the testfile as the last argument?

Comment (tautschnig) [inline regression/goto-instrument-unwind/unwind-chain.sh]: So maybe test.pl should be tweaked to simplify such scripts? Else there could be a brief options parser in this code to extract the last argument from $@.

**Comment by danpoe** [comment]
Ok, I think I've addressed everything.

File: src/util/string_utils.cpp [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+
+  Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+// unit tests
```
Comment (tautschnig) [inline src/util/string_utils.cpp]: I would suggest to place this in the unit/ top-level folder.

File: src/goto-instrument/goto_instrument_parse_options.cpp [hunk 16]
```diff
@@ -147,15 +145,90 @@ int goto_instrument_parse_optionst::doit()
...
 
-      goto_unwind(goto_functions, k);
+      if(unwind || unwindset || unwindset_file)
+      {
+        int k=-1;
 
-      goto_functions.update();
-      goto_functions.compute_loop_numbers();
+        if(unwind)
+          k=stoi(cmdline.get_value("unwind"));
```
Comment (tautschnig) [inline src/goto-instrument/goto_instrument_parse_options.cpp]: I'm a bit surprised this works without std::, but it has done before. Maybe use safe_string2int so that we can later put in std::stoi across the codebase?

File: .travis.yml [hunk 17]
```diff
@@ -1,22 +1,43 @@
 language: cpp
 
-os:
```
Comment (tautschnig) [inline .travis.yml]: Did you intentionally add .travis.yml changes to this pull request? (I am in favour of those changes, just asking.)

File: src/util/string_utils.cpp (hunk 15 above)
Comment (danpoe) [inline src/util/string_utils.cpp]: Ok, I've added this to the unit tests folder.

File: .travis.yml (hunk 17 above)
Comment (tautschnig) [inline .travis.yml]: Ok, thanks for clarifying!

**Comment by tautschnig** [comment]
Unless anyone else voices concerns, it should just be a rebase that's blocking a merge.

File: src/goto-instrument/goto_instrument_parse_options.cpp [hunk 18]
```diff
@@ -147,15 +145,90 @@ int goto_instrument_parse_optionst::doit()
...
+        if(unwind)
+          k=safe_string2int (cmdline.get_value("unwind"));
+
+        unwind_sett unwind_set;
+
+        if(unwindset_file)
+        {
+          std::string us;
+          std::string fn=cmdline.get_value("unwindset-file");
+          std::ifstream file(fn);
```
Comment (kroening) [inline src/goto-instrument/goto_instrument_parse_options.cpp]: This unfortunately needs to include widen() on WIN32, or Unicode filenames won't work.

File: src/util/string_utils.cpp [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+    if(!std::isspace(s[i]))
+      break;
+  }
+  if(i==n)
+    return "";
+
+  // find last non-space char
+  long j; // need signed type here
+  for(j=n-1; j>=0; j--)
+  {
```
Comment (danpoe) [inline src/util/string_utils.cpp]: I've used a reverse iterator for this now.

**Comment by danpoe** [comment]
This is now rebased and I've addressed the additional comments.
